homeassistant/components/mqtt/__init__.py

```diff
@@ -707,26 +712,24 @@ def async_remove() -> None:
 
         return async_remove
 
-    async def _async_unsubscribe(self, topic: str) -> None:
+    @callback
+    def _async_unsubscribe(self, topic: str) -> None:
         """Unsubscribe from a topic.
 
         This method is a coroutine.
         """
-        async with self._paho_lock:
-            result = None  # type: int
-            result, _ = await self.hass.async_add_job(
-                self._mqttc.unsubscribe, topic)
-            _raise_on_error(result)
+        result = None  # type: int
+        result, _ = self._mqttc.unsubscribe(topic)
+        _raise_on_error(result)
 
-    async def _async_perform_subscription(self, topic: str, qos: int) -> None:
+    @callback
+    def _async_perform_subscription(self, topic: str, qos: int) -> None:
         """Perform a paho-mqtt subscription."""
         _LOGGER.debug("Subscribing to %s", topic)
 
-        async with self._paho_lock:
-            result = None  # type: int
-            result, _ = await self.hass.async_add_job(
-                self._mqttc.subscribe, topic, qos)
-            _raise_on_error(result)
+        result = None  # type: int
+        result, _ = self._mqttc.subscribe(topic, qos)
+        _raise_on_error(result)
 
     def _mqtt_on_connect(
```

Comment:
This will also be called in the event loop. So remove add job here.

Ans also for `_mqtt_on_message`.

Reply:
Ok. It's a bit weird that they consider this ok to use with asyncio while having all those locks. 🤔